---
title: "PBMC atac analysis using Signac"
output: html_document
---

# 1. Pre-processing

```{r}
library(Signac)
library(Seurat)
library(GenomicRanges)
library(ggplot2)
library(patchwork)
```
```{r}
setwd("~/Desktop/Python/GitHub/scRNA-Scanpy-DataPlayground/PBMC10k")
```

```{r}
# Load your Seurat object
atac <- readRDS("atac_qc_filtered.rds")
```

```{r}
atac
```

```{r}
head(atac, 5)
```

```{r}
#get the Peak/Cell matrix (row= peak, column = cell)
counts_matrix <- atac[["RNA"]]$counts
#head(counts_matrix, 3)
```

```{r}
#get metadata
metadata <- atac@meta.data
head(metadata, 5)
```

```{r}
#load fragment files
frag_path <- "~/Desktop/Python/GitHub/scRNA-Scanpy-DataPlayground/PBMC10k/pbmc10k/atac_fragments.tsv.gz"
```

```{r}
library(BiocFileCache)
library(AnnotationHub)
library(ensembldb)
```

### annotation
```{r}
ah <- AnnotationHub()
# Search for the Ensembl 98 EnsDb for Homo sapiens on AnnotationHub
query(ah, "EnsDb.Hsapiens.v98")
```

```{r}
ensdb_v98 <- ah[["AH75011"]]
```

```{r}
# extract gene annotations from EnsDb
annotations <- GetGRangesFromEnsDb(ensdb = ensdb_v98)
```
```{r}
# Convert to UCSC style
seqlevelsStyle(annotations) <- "UCSC"
```

```{r}
annotations
```

```{r}
chrom_assay <- CreateChromatinAssay(counts = counts_matrix,
                                    sep = c(":", "-"),
                                    fragments = frag_path,
                                    min.cells = 10,
                                    min.features = 200)
```

```{r}
pbmc <- CreateSeuratObject(counts = chrom_assay,
                           assay = "peaks",
                           meta.data = metadata)
```

```{r}
pbmc[['peaks']]
```
```{r}
#add gene information to the object
Annotation(pbmc) <- annotations
```

```{r}
pbmc[['peaks']]
```

```{r}
granges(pbmc)
```
```{r}
#remove the features that correspond to chromosome scaffolds e.g. (KI270713.1) or other sequences instead of the (22+2) standard chromosomes.
peaks.keep <- seqnames(granges(pbmc)) %in% standardChromosomes(granges(pbmc))
pbmc <- pbmc[as.vector(peaks.keep), ]
```

```{r}
granges(pbmc)
```

# 2. QC metrics
## 2.1. Nucleosome banding pattern (nucleosome_signal)
```{r}
# compute nucleosome signal score per cell
pbmc <- NucleosomeSignal(object = pbmc)
```

## 2.2. Transcriptional start site (TSS) enrichment score
```{r}
# compute TSS enrichment score per cell
pbmc <- TSSEnrichment(object = pbmc)
```
```{r}
#check metadata
pbmc@meta.data
```

## 2.3. Total number of fragments in peaks
#we have column "nCount_peaks"

## 2.4. Fraction of fragments in peaks
```{r}
#pbmc$pct_reads_in_peaks <- pbmc$peak_region_fragments / pbmc$passed_filters * 100
#I cannot calculate this QC matric because I don't have the column "passed_filters"
```

## 2.5. Ratio reads in genomic blacklist regions

```{r}
pbmc$blacklist_ratio <- FractionCountsInRegion(
  object = pbmc, 
  assay = 'peaks',
  regions = blacklist_hg38_unified
)
```

```{r}
#check metadata
pbmc@meta.data
```

